mlir/lib/Bindings/Python/IRModule.cpp

```diff
@@ -57,15 +57,18 @@ void PyGlobals::loadDialectModule(llvm::StringRef dialectNamespace) {
     break;
   }
 
+  if (loaded.is_none())
+    return false;
   // Note: Iterator cannot be shared from prior to loading, since re-entrancy
   // may have occurred, which may do anything.
-  loadedDialectModulesCache.insert(dialectNamespace);
+  loadedDialectModules.insert(dialectNamespace);
+  return true;
```

Comment:
I think the negative caching that used to happen here actually made sense (avoiding python interpreter work seems significant vs just C++ map lookups that the other caching potentially avoided)

Reply:
I don't see any previous negative caching at module level, am I missing something?

It could make sense to add _here_. Several lookup functions are calling this, and in absence of the module for a specific dialect prefix, this will keep hitting the slow path of trying to load a module from a filesystem. This seems undesirable unless module import has its own cache. (I generally prefer adding optimizations when we know they are actually meaningful...) We should document this behavior and provide a way to reset the (negative) cache programmatically for extreme cases like the module being placed into one of the search paths by another process.

Reply:
> I don't see any previous negative caching at module level, am I missing something?

The negative cache currently works in that when a dialect is "looked up" but fails to have its module found (for lack of existence or lack of prefix search path) will still be `loadedDialectModulesCache.insert(dialectNamespace);` (the loop just finishes silently).

> Several lookup functions are calling this

That's true but there's only one (under reasonable circumstances) that could possibly actually trigger a load that doesn't succeed (as far as I can imagine): `lookupOperationClass` during `Operation.create` and only if you're creating an operation with/from an unregistered dialect.

> absence of the module for a specific dialect prefix

This is what the added test code below exercises, demonstrates (although obliquely). In the absense of the added `custom_dialect.custom.py`, it's exactly as you say (the lookup happens over and over) but with the addition of that module it only happens once.

Reply:
Thanks for the explanation!

> That's true but there's only one (under reasonable circumstances) that could possibly actually trigger a load that doesn't succeed (as far as I can imagine)

If we have a type/attr from an unregisterd/unloadable dialect, I believe we can still process that opaquely and have a python object representing that type/attr. Wouldn't obtaining such an object trigger caster lookup that always lead to repeated module searches?

> In the absense of the added custom_dialect.custom.py, it's exactly as you say (the lookup happens over and over) but with the addition of that module it only happens once.

I suppose creating ops from an unloaded/unregistered dialect isn't very common, so it's not critical to improve that, but still nice to have.

Reply:
Yea so I guess it's true you can get into these weird corners where you want to do stuff in python but you haven't setup any of the boiler plate? I agree that I don't think it's common and adding "cache invalidation" mechanisms, such as we have currently (and this PR is meant to excise) to `loadedDialectModules` probably isn't worth it.

Reply:
Is there any follow up needed here? E.g., adding a TODO or github issue to track? (It sounds like agreement that this is only bad in what should be a very uncommon case and the added complexity/footgun doesn't warrant it).